refactor: use unity.mathematics.half in place of added types

NoelStephensUnity opened the pull request:
Reducing the footprint of the Unity.Mathematics types due to the vast amount of added public API changes required.
This should have minimal impact on the API footprint.

## Testing and Documentation

- No tests have been added.
- Includes minor modification to existing integration tests.
- No documentation changes or additions were necessary.

NoelStephensUnity committed:
update

adding asmdef references to Unity.Mathematics

NoelStephensUnity committed:
refactor

updated HalfVector3 and HalfVector4 with Unity.Mathematics types.
removed classes no longer needed.
renamed HalfVector3DeltaPosition to NetworkDeltaPosition and made that use HalfVector3.

NoelStephensUnity committed:
style

removing whitespace

NoelStephensUnity committed:
update

setting the Vector4 to size of 4 (not 3).
Updating the teleport fix for testing purposes.

NoelStephensUnity committed:
test

mirroring PR-2451 changes to NetworkTransformStateTests for the teleport fix.

NoelStephensUnity committed:
revert

Reverting out the serialization pipeline updates

NoelStephensUnity committed:
update

A simpler approach to adding Unity.Mathematics.Half without adding any additional runtime public APIs.

0xFA11 commented on com.unity.netcode.gameobjects/Components/HalfVector3.cs:
```diff
@@ -13,45 +14,70 @@ namespace Unity.Netcode.Components
     /// </remarks>
     public struct HalfVector3 : INetworkSerializable
     {
+        public const int Size = 3;
```
Is this `Size` or `Length`?

`Size` kind of implies size in the memory or on the network, usually in bytes but it's not it right?
`Length` is implying "how many items does this container contain?" and that's 3 items — and 3 items x 2 bytes = 6 bytes in total.

Maybe a nit-pick but wanted to bring it up (same applies to `HalfVector4` too)

0xFA11 replied:
or in general, maybe never expose this constant value to public since we're using it internally for the `for` loops and 3 for HalfVector3 and 4 for HalfVector4 are kind of obvious by their name — what do you think?

NoelStephensUnity replied:
Good call out. That needed to be non-public and length does indeed make more sense then size.

NoelStephensUnity committed:
update

Implementing Fatih's suggestion which is to rename size to length and non-public.
